compiler/rustc_codegen_llvm/src/builder.rs

```diff
@@ -669,81 +670,36 @@ impl BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
-        // WebAssembly has saturating floating point to integer casts if the
-        // `nontrapping-fptoint` target feature is activated. We'll use those if
-        // they are available.
-        if self.sess().target.arch == "wasm32"
-            && self.sess().target_features.contains(&sym::nontrapping_dash_fptoint)
-        {
+        if llvm_util::get_version() >= (12, 0, 0) {
             let src_ty = self.cx.val_ty(val);
             let float_width = self.cx.float_width(src_ty);
             let int_width = self.cx.int_width(dest_ty);
-            let name = match (int_width, float_width) {
-                (32, 32) => Some("llvm.wasm.trunc.saturate.signed.i32.f32"),
-                (32, 64) => Some("llvm.wasm.trunc.saturate.signed.i32.f64"),
-                (64, 32) => Some("llvm.wasm.trunc.saturate.signed.i64.f32"),
-                (64, 64) => Some("llvm.wasm.trunc.saturate.signed.i64.f64"),
-                _ => None,
-            };
-            if let Some(name) = name {
-                let intrinsic = self.get_intrinsic(name);
-                return Some(self.call(intrinsic, &[val], None));
-            }
+            let name = format!("llvm.fptosi.sat.i{}.f{}", int_width, float_width);
+            let intrinsic = self.get_intrinsic(&name);
+            return Some(self.call(intrinsic, &[val], None));
         }
-        None
-    }
 
-    fn fptosui_may_trap(&self, val: &'ll Value, dest_ty: &'ll Type) -> bool {
-        // Most of the time we'll be generating the `fptosi` or `fptoui`
-        // instruction for floating-point-to-integer conversions. These
-        // instructions by definition in LLVM do not trap. For the WebAssembly
-        // target, however, we'll lower in some cases to intrinsic calls instead
-        // which may trap. If we detect that this is a situation where we'll be
-        // using the intrinsics then we report that the call map trap, which
-        // callers might need to handle.
-        if !self.wasm_and_missing_nontrapping_fptoint() {
-            return false;
-        }
-        let src_ty = self.cx.val_ty(val);
-        let float_width = self.cx.float_width(src_ty);
-        let int_width = self.cx.int_width(dest_ty);
-        matches!((int_width, float_width), (32, 32) | (32, 64) | (64, 32) | (64, 64))
+        None
     }
 
     fn fptoui(&mut self, val: &'ll Value, dest_ty: &'ll Type) -> &'ll Value {
-        // When we can, use the native wasm intrinsics which have tighter
-        // codegen. Note that this has a semantic difference in that the
-        // intrinsic can trap whereas `fptoui` never traps. That difference,
-        // however, is handled by `fptosui_may_trap` above.
-        //
-        // Note that we skip the wasm intrinsics for vector types where `fptoui`
-        // must be used instead.
-        if self.wasm_and_missing_nontrapping_fptoint() {
+        // This intrinsic is only used with non-saturating casts that have UB on
```

Comment:
Nit: this is not really an intrinsic, but rather a builder method/instruction builder/etc?

From this comment it is no longer clear why it is important that we specifically go through an effort  on wasm to pick a behaviour for what's otherwise an undefined behaviour. It would be good to keep the remark about the intrinsics producing better machine code.

(as a side note: From spelunking it seems that `simd_cast` also uses this (cc @workingjubilee @calebzulawski) which is potentially undesired (similar to the issue fixed in https://github.com/rust-lang/rust/pull/84274). This implementation would also crash in a simd setting on wasm, but that's a future improvement)

Reply:
It does seem a little odd to trap on UB, my understanding is that this function is really only for `{f32,f64}::to_int_unchecked` which assumes the caller already checked the bounds.

This doesn't crash on wasm simd (we test the simd version of `to_int_unchecked` on wasm)--it looks like it falls back to the generic llvm instruction which supports vectors

Reply:
I've changed "intrinsic" to "method", and I've elaborated on the comment to hopefully help explain why wasm is different here.

This should work for SIMD types since it specifically avoids (a few lines below) any input types that are vectors.